core/storage/opportunity/gae_models.py

```diff
@@ -208,7 +208,8 @@ def get_user_id_migration_policy():
         return base_models.USER_ID_MIGRATION_POLICY.NOT_APPLICABLE
 
     @classmethod
-    def get_skill_opportunities(cls, page_size, urlsafe_start_cursor):
+    def get_skill_opportunities(
+            cls, page_size, urlsafe_start_cursor, shouldFetchAll=False):
```

Comment:
Should use snake_case for args (please follow surrounding code).

I'm somewhat concerned about this, though. It seems to conflict with page_size and is different from how we handle other pagination methods in the codebase. Can we not have this option?

(Also, I would have expected this code to come with backend tests for it.)

Reply:
I see. Would a separate method for retrieving all entities suffice (and reverting the changes to this method)? And I will definitely address the other issues as well.

Reply:
Could you explain why there is a need to "retrieve all entities"? This seems like it would be a problem generally unless the number of entities is somehow capped (is it?).

Reply:
To support search functionality of the skill opportunities in the frontend, we need to retrieve all skill opportunities, join them with their corresponding topics for topic information, and then filter based on the skill description and/or topic name. We're not able to directly use a filter at the storage layer, since the SkillOpportunity model doesn't have topic information.

Reply:
But that won't scale, right? There could be a lot of skill opportunities, so doing that computation every time seems very resource-intensive. I think we'll need to do something where we actually precompute and store that information for easy retrieval.

Also, have you looked into the Search API? We use that currently in the library page and you can find more info in core/domain/search_services.py. I'm not sure whether it solves your use case though, it's just an alternative to look into.

Reply:
Hmm, any idea what that pre-computation would look like? Or is there any precedent in the current codebase? The search query is only made upon pressing the enter key, so it's at least not making a full fetch every key stroke.

And no, I wasn't aware of the Search API. I will take a look, thanks!

Reply:
Can you write a short [design doc](https://github.com/oppia/oppia/wiki/Writing-design-docs) with the problem statement, and share it? I think the design depends on what requirements you want to support (e.g. do you want matches on prefixes, etc.)

In general, implementing search is not straightforward, we need to design it carefully. You might want to move that part to a separate PR instead -- it could very easily turn into its own big project.

Reply:
Sure! I split up the search capability into its own issue [here](https://github.com/oppia/oppia/issues/8559). I'll write a design doc for the search feature and update the issue accordingly.

For now, I've reset the search feature to an older version (where there are no backend changes) and hidden the search bar in the frontend via a flag.